### ISSUE TITLE
Integers cannot be passed as Bind Parameter on Python 3 (MS Access)

### ISSUE THREAD

**Issue by schlamar** [mkleehammer/pyodbc#84]
See this SA question for details: http://stackoverflow.com/q/20240130

I can reproduce this issue on Windows 10, Python 3.5 32bit, pyodbc 3.0.10. Converting integers to floats before passing them to `execute` is a work around (as stated in the second answer in the SA question).

**Comment by mkleehammer** [comment]
FYI: Not an issue with Python 3.5.0 and pyobc 3.0.10 on OS X with PostgreSQL.

Some information about the table would be helpful.  The *most* helpful would be a SQL trace, particularly of a failed pyodbc attempt and a successful pypyodbc attempt to see how they are binding differently.

**Comment by mkleehammer** [comment]
I'm working on a Python 3.5.1 test.

**Comment by mkleehammer** [comment]
Not able to reproduce with SQL Server, though this is a 64-bit test.  I'll test 32 after I install Python 3.5.1 32-bit.

```text
python:  3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:54:25) [MSC v.1900 64 bit (AMD64)]
pyodbc:  3.0.10 C:\dev\pyodbc\build\lib.win-amd64-3.5\pyodbc.cp35-win_amd64.pyd
odbc:    03.80.0000
driver:  sqlncli11.dll 11.00.3000
         supports ODBC version 03.80
os:      Windows
unicode: Py_Unicode=2 SQLWCHAR=2
         7 6.1.7601 SP1 Multiprocessor Free
----------------------------------------------------------------------
Ran 190 tests in 0.906s
```

I'm thinking this is an Access-specific problem.  I'm going to update the title to reflect this.  Can anyone provide small example (access file included) to reproduce?

**Comment by schlamar** [comment]
Attached a test case with an access database:
[Testcase.zip](https://github.com/mkleehammer/pyodbc/files/147796/Testcase.zip)

**Comment by smsearcy** [comment]
I'm encountering what seems to be the same issue but a different database (Marxmeier SQL/R ODBC driver).  I'm using pyodbc 3.0.10 w/ Python 3.4.3 (32-bit) on Windows 8.1 (I do not have the error with Python 2.7.11).

When trying to pass an integer as a parameter to query it raises the following error:

    pyodbc.Error: ('HYC00', '[HYC00] [Microsoft][ODBC Driver Manager] Driver does not support this parameter (0) (SQLBindParameter)')

Similar to the Stacked Overflow discussion linked above it works if I cast the parameter as a `float`.

Here is some more information about the column in case that is useful (I'm not sure what sort of traces I'm able to pull with with this ODBC driver):

row.cursor_description:
`('column', <class 'int'>, None, 11, 11, 0, True)`
cursor.columns:
`('dsn', 'schema', 'table', 'column', 4, 'INTEGER', 11, 4, 0, None, 1, None, None, 4, None, None, 1, 'YES')`

If this isn't relevant to this issue I can create a separate issue.

**Comment by smsearcy** [comment]
Update, I'm having the same issue still on Python 3.5.2 64-bit so it doesn't appear to be a 32-bit vs 64-bit issue (at least for Python 3).

Looking at `params.cpp` it seems this is probably related to Python 3 dropping the distinction between int and long.  According to [Microsoft Access Data Types](https://msdn.microsoft.com/en-us/library/ms714540(v=vs.85).aspx) it appears that Access does not support SQL_BIGINT.

So is there a way in `GetLongInfo()` to fallback to SQL_INTEGER if the driver does not support big integers?

I'm going to test what happens in Python 2.7 with long integers and/or 64-bit.

**Comment by gordthompson** [comment]
I found this thread while searching after seeing a recent Stack Overflow question.

I can reproduce the issue using the above Testcase.zip with pyodbc 4.0.0 under Python 3.5.2 (64-bit on Windows) using the Access 2010 ODBC driver.

I have made fairly extensive use of pyodbc with Access ODBC under Python 2.7, so I strongly suspect that this issue is related to Python 3.x.

**Comment by gordthompson** [comment]
Confirmed still an issue with pyodbc 4.0.3 under

> Python 3.4.4 (v3.4.4:737efcadf5a6, Dec 20 2015, 19:28:18) [MSC v.1600 32 bit (Intel)] on win32

On the same machine, the exact same test code (from Testcase.zip, above), pyodbc version , and ODBC driver work fine when using 

> Python 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)] on win32

**Comment by smsearcy** [comment]
@gordthompson, I think those logs support my theory that the issue is in Python 3 all integers are longs so they are being bound to `SQL_BIGINT` which isn't supported by Access.

It seems that there needs to be a condition in `GetLongInfo()` that will set `info.ParameterType` to `SQL_INTEGER` if the driver does not support `SQL_BIGINT` or (perhaps simpler) if the number is small enough to fit in an `SQL_INTEGER`.

I have not written any C/C++ in over 10 years but I might try my hand at a fix.

**Comment by smsearcy** [comment]
Thank you, integer parameters are now working correctly with the ODBC driver I am using.

**Comment by acdr000** [comment]
FWIW, I am encountering this issue with Python 2.7.10.2. Casting all integer parameters to floats indeed provides a work-around.

**Comment by gordthompson** [comment]
@acdr000 - This particular issue was related to Python_3, not Python_2. Please open a new issue and provide relevant details as described in the issue template (platform, Python version, pyodbc version, ODBC driver and version, etc.).

**Comment by SpringsTea** [comment]
I'm getting this error with python 3.7.4
Had to rollback to 2.7

**Comment by gordthompson** [comment]
@SpringsTea - This issue was fixed in 4.0.8 and then broken again in 4.0.25. See #509 for the current status.
